### PR TITLE
[instrument_builder] Add REDCap CSV --> .linst functionality

### PR DESCRIPTION
**Pull request by zaliqarosli**
## Brief summary of changes

- [x] Have you updated related documentation?

This PR adds the functionality to load onto the instrument builder a REDCap data dictionary for an instrument in .csv format. Note: you cannot load the entire REDCap data dictionary - you have to extract a single instrument's data dictionary for upload.

The loading functionality will render the instrument under the 'Build' tab as usual i.e. as for linst format, and allow to 'Save' the data dictionary as a .linst instrument. I have also added the capability of downloading rules if they exist in the .csv file.

_**Caveat:**_ This PR is dependent on a new XINValidate feature. I created a "CONTAINS" operator for the XINValidateMultiSelect() function of NDB_BVL_Instrument_LINST.class.inc (on QPN), to work with REDCap rules for multiselect. E.g.:

```
[site] = '3' <--- normal select element
[income_source(8)] = '1' <--- multi select i.e. this field is required if option 8 was selected (among other potential options selected)
```
I can either add the "CONTAINS" feature in this PR, in a separate PR, or take it out completely as it doesn't yet exist in LORIS.

**_Another note_**: I had to play with already messy code, some that were there but never used i.e. `modules/instrument_builder/js/instrument_builder.rules.js ` seems to be doing something but was never implemented on the module. I tried to utilize it for this PR, but maybe there is a better way to code the .rules download. I'm open to all feedback, I coded this all fairly quickly.

#### Testing instructions (if applicable)

1. Load a test .linst file as you would normally. Notice how the instrument is rendered on the 'Build' tab. Make some modifications to the instrument if you like. Save the instrument under the 'Save' tab. Look at the linst file and make sure it is what you expect.
2. Take a REDCap .csv data dictionary for 1 instrument. (If you don't have one, let me know and I can share one with you). Load the .csv file as you would a .linst file. The instrument should be rendered as you would expect in the 'Build' tab. Make sure there are no errors. You can 'Save' the instrument, giving you a .linst file and maybe a .rules file. Read the .linst  file and make sure it is what you expect, following the same format as the linst file generated from step 1.
3. Generate the SQL file for the instrument: `cat $instrument.linst  | php tools/generate_tables_sql_and_testNames.php`
Check the SQL statements were generated fine.
4. Run the SQL script and make sure you get no SQL errors.
5. Add the instrument to the test_battery table.
6. Run `php tools/assign_missing_instruments.php`
7. Test the instrument on the front end, and make sure it works like any normal linst instrument

#### Link(s) to related issue(s)

* Resolves #  (Reference the issue this fixes, if any.)

### CODE REVIEW

File: modules/instrument_builder/js/instrument_builder.instrument.js [hunk 1]
```diff
@@ -357,9 +358,226 @@ var Instrument = {
...
+                  if (branching != undefined && branching != '') {
+                    // then add rules
+                    let dependentValue = '';
+                    let dependentField = branching.match(/\[(.*?)\]/) ? branching.match(/\[(.*?)\]/)[1] : '';
+                    if (dependentField.includes("(") && dependentField.includes(")")) {
+                      // multiselect with CONTAINS
+                      let substring = dependentField.match(/\((.*?)\)/);
+                      dependentValue = substring[1];
+                      dependentField = dependentField.replace(substring, '');
+                      rulesContent += elementName + "{-}Required{-}" + dependentField + "{@}CONTAINS{@}" + dependentValue + "\n";
```
Comment (zaliqarosli) [inline modules/instrument_builder/js/instrument_builder.instrument.js:510]: need to add "CONTAINS" functionality to XIN rules

File: modules/instrument_builder/jsx/react.instrument_builder.js [hunk 2]
```diff
@@ -126,6 +126,15 @@ class LoadPane extends Component {
     return (
       <TabPane Title='Load Instrument' {...this.props}>
         <div className='col-sm-6 col-xs-12'>
+          <i>
+            Note: You may load either a LORIS .linst format or a REDCap
+            .csv format.
+          </i><br></br>
+          <i>
+            If loading a .csv format, make sure the file contains the data
```
Comment (zaliqarosli) [inline modules/instrument_builder/jsx/react.instrument_builder.js:134]: caveat: you can only load 1 instrument data dictionary at a time

File: modules/instrument_builder/js/instrument_builder.instrument.js [hunk 3]
```diff
@@ -357,9 +358,226 @@ var Instrument = {
...
+                    lineContent += "static{@}";
+                    lineContent += elementName + "{@}" + elementDescription + "\n";
+                    elementName = '';
+                    break;
+                  case "descriptive":
+                    lineContent += "static{@}{@}" + elementDescription + "\n";
+                    elementName = '';
+                    break;
+                }
+                // Add rules
```
Comment (zaliqarosli) [inline modules/instrument_builder/js/instrument_builder.instrument.js:499]: added ability to read branching logic in csv and download .rules file along side .linst file

**Comment by jesscall** [comment]
Tested this PR with a RedCAP .csv data dictionary provided by @christinerogers and it doesn't seem to be compatible.


Screen Shot 2021-09-27 at 11 48 21 AM
Screen Shot 2021-09-27 at 11 48 34 AM

Not all fields name are detected, the radio button format seems off, the html is not parsed and so things like tables are not translated well 

@zaliqarosli, maybe you could provide an example of the .csv you use for CONP?

**Comment by zaliqarosli** [comment]
@driusan i think we can go ahead and remove this from the list of critical to release - looks like theres a few bugs that need to be worked out

**Comment by zaliqarosli** [comment]
replaced by #8282

**Pull request closed without merging**